PR: Set projet filenames correctly when closing a project

jnsebgosselin opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

<!--- Explain what you've done and why --->

When closing a project, call `set_project_filenames` before setting the current project to `None`.


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #8375


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: Jean-Sébastien Gosselin

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @jnsebgosselin! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`spyder/plugins/projects/plugin.py`](https://github.com/jnsebgosselin/spyder/blob/78e421dade25801ba5ce797c4ad28800a337cb82/spyder/plugins/projects/plugin.py) !

jnsebgosselin commented:
@ccordoba12 Is this worth it to add a test for such a small fix? Would you?

jnsebgosselin commented:
Forget my last comment, I found a way to cover this at the plugin level without too much effort.

CAM-Gerlach commented on spyder/plugins/projects/tests/test_plugin.py:
```diff
@@ -120,5 +127,33 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
+
+    # We mock os.path.isfile so that we do not have to
+    # actually create the files on the disk.
+    mocker.patch('spyder.plugins.projects.api.os.path.isfile',
```
Is this what we're using now instead of ``monkeypatch``?

CAM-Gerlach replied:
I don't know, that was just my understanding from the previous files but its up to @ccordoba12

ccordoba12 replied:
I don't have a problem with using any approach.

CAM-Gerlach commented on spyder/plugins/projects/tests/test_plugin.py:
```diff
@@ -25,16 +25,25 @@
-# =============================================================================
-# Fixtures
-# =============================================================================
+# ---- Fixtures
```
??? I thought this was our standard test file formatting for new and substantially revised files? I've never seen the ``# ---- NAME`` formatting anywhere else in our code (or anywhere else that I've seen) while it seems the previous is much more conventional (and besides, looks way better and clearer IMO).

ccordoba12 replied:
I agree with @CAM-Gerlach here. Please revert.

jnsebgosselin replied:
!!! The reason I did this was because special comments are added to the Outline Explorer, which I find very useful to separate at a glance the fixtures from the tests. But that's ok, I'm going to revert it.

![image](https://user-images.githubusercontent.com/10170372/49959781-8efcea00-fedc-11e8-9600-32a15370f6b0.png)

ccordoba12 replied:
Sorry, then I agree with your change. Please remove your last commit.

CAM-Gerlach replied:
What about

```python
# =============================================================================
# ---- Fixtures
# =============================================================================
```

That keeps most of the existing convention (and more visible/better looking styling) and only requires adding four characters on one line to existing files as they are modified, and is more visible inline without the outline explorer while still achieving the same affect with it.

jnsebgosselin replied:
👍 👍 👍 I think this is a good idea. If @ccordoba12 agrees with it I will do that.

ccordoba12 replied:
Sure, I agree with that too.

CAM-Gerlach commented on spyder/plugins/projects/tests/test_plugin.py:
```diff
@@ -57,9 +66,7 @@ def __getattr__(self, attr):
-# =============================================================================
-# Tests
-# =============================================================================
+# ---- Tests
```
Same as above

CAM-Gerlach commented on spyder/plugins/projects/tests/test_plugin.py:
```diff
@@ -120,5 +127,33 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
+    assert projects.get_project_filenames() == []
+
+    # Close and reopen the project.
+    projects.close_project()
```
Although the original bug only occurred on closing and re-opening projects, would it be too much work to test switching projects as well in case there's a regression or edge case in the future?

ccordoba12 replied:
I agree with this too. It shouldn't be that hard.

jnsebgosselin replied:
This is a good idea. I will add that.

ccordoba12 commented:
@jnsebgosselin, you need to rebase/merge with master to fix the tests in Azure.

jnsebgosselin commented:
@ccordoba12 I added the requested additional test. There is a chance you won't like my `create_projects` fixture that I've created to remove some repetitiveness of the code in both tests I've added.

There is no problem if you don't like it, I will just remove the 2 latest commits.

ccordoba12 commented on spyder/plugins/projects/tests/test_plugin.py:
```diff
@@ -120,5 +156,49 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
+
+    # Create the projects plugin.
+    projects = create_projects(path, opened_files)
+    assert projects.get_project_filenames() == []
```
Why is this empty and not equal to `opened_files`?

jnsebgosselin replied:
`get_project_filenames` returns what is saved in the project's config file and this gets updated only when closing the project, not as new files are opened or closed in the Editor.

jnsebgosselin replied:
So `create_project` open a new project with an empty list of files at first in the config. Then when we close the project, the list of files we are mocking as open in the Editor are set as the project filenames in the config file.

ccordoba12 replied:
Ok, I see. Thanks for the explanation!
